org.eclipse.xtext.xtext.generator/src/org/eclipse/xtext/xtext/generator/grammarAccess/GrammarAccessExtensions.xtend

```diff
@@ -396,8 +391,61 @@ class GrammarAccessExtensions {
 		}
 	}
 	
-	def String grammarFragmentToString(EObject ele, String prefix) {
-		return serializer.grammarFragmentToString(ele, prefix)
+	def String grammarFragmentToString(EObject object, String prefix) {
+		val node = NodeModelUtils.findActualNodeFor(object)
+		if (node === null) {
+			if (object instanceof RuleCall)
+				if (object?.rule?.name !== null)
+					return process(object.rule.name, prefix)
+			return ""
+		} else {
+			return node.text.process(prefix)
+		}
+	}
+
+	private def process(String input, String prefix) {
+		// remove leading and trailing blank lines
+		var lines = input.split('(\\r?\\n)')
+		var first = 0
+		while (isBlank(lines.get(first)))
+			first++
+		var last = lines.length - 1
+		while (isBlank(lines.get(last)))
+			last--
+		lines = lines.subList(first, last + 1)
+		// remove common whitespace (e.g. leading blanks) and add prefix to each line
+		val commonWhitespace = commonLeadingWhitespace(lines)
+		for (var n=0; n<lines.length; n++)
+			lines.set(n, prefix + lines.get(n).substring(commonWhitespace.length))
+		// generate result by joining lines
+		return lines.join(System.lineSeparator)
+	}
+
+	private def isBlank(String line) {
+		return line.trim.empty
+	}
+	
+	private def String commonLeadingWhitespace(List<String> lines) {
```

Comment:
`com.google.common.base.Strings.commonPrefix(CharSequence, CharSequence)` may be handy here, too along with `isBlank`

Reply:
Have thought about this a bit. As the existing method computes the prefix for only two strings I would have to iterate all the lines, compute the common prefix with the first line, then take the shortest prefix, remove all trailing non whitespaces, check that is blank and then return it. Do not believe it will be a noticeable improvement.

Reply:
I didn't test this, but my gut feeling is, that a normal loop along with the utility method is also very readable:
```
String commonLeadingWS(List<String> lines) {
  if (lines.size() < 2) return "";
  // start at second line since the first is left-trimmed in any case
  String current = Strings.repeat(" ", tabsToSpaces(lines.get(1)).length())
  for(int i = 1; i < lines.length; i++) {
    String next = lines.get(i);
    if (!isBlank(next))) {
      current = Strings.commonPrefix(current, tabsToSpaces(lines.get(i));
    }
  }
  return current;
}
```

Reply:
Get your idea and guess it will work (we could NOT skip the first element but that is minor).

Modified it a bit to the following:

```
	private def String commonLeadingWhitespace(List<String> lines) {
		if(lines.size() < 2) return "";
		var current = Strings.repeat(" ", lines.get(0).replaceAll("\t", "    ").length())
		for (var i = 0; i < lines.length; i++) {
			val next = lines.get(i).replaceAll("\t", "    ");
			if (!isBlank(next))
				current = Strings.commonPrefix(current, next);
		}
		return current;
	}
```

Do you think a replacement of a tab with 4 spaces will work well? Or do we have to read some preferences or something like that? If the user uses tabs and spaces inconsistenly and has one tab for e.g. 2 spaces, the resulting formatting might be wrong ...

Reply:
It was only an idea. We could use common prefix without that replace operation.

Reply:
Why I excluded the first line from the common prefix:
```
SomeGrammarRule: 
  {NewThingy} ('{'
    elements+=Element;
  '}');
```

When you look at the group in parentheses, you'd want to strip the leading space from the last two lines but the element doesn't have a leading space in the first line. I didn't check the current approach carefully though it might indeed be flawed to exclude line zero.

Reply:
Ok, you are right. In case of inconsistent usage of blanks and tabs we will get wired results in any case. With or without the replace operation. In any case your solution is much more readable.

Reply:
Because of the skipping of the first line ... maybe I get you wrong, but in your example we do not want to remove any leading whitespace ... because of the first line. In this sense it is the counter example ... we need to consider the first line.

Reply:
The grammar rule has a body of type `Group` - the group has two elements: one action and another group .
The nested group shall be put into the grammar access as a comment. It starts at the opening parenthesis and the keyword: `(`{``. So the raw text of that group is 
```
('{'
    elements+=Element;
  '}');
```
Note the indentation of the last two lines which is off by two chars. If the first line considered in the leading WS computation, there won't be any whitespace to be trimmed. That was my line of thought.

Reply:
I see. In that case the algorithm will not work. On the other hand the complete rule will be misformatted in case we skip the first line. Guess it is not possible to make every case 100% "correct" ... hence we have to try to make the most common cases good looking. And the complete rule is probably the most important case.